first_order_model/shrink_util.py

```diff
@@ -0,0 +1,1198 @@
+    """
+    #    _ = model(inp)
+    driving_lr = x.get('driving_lr', None)
+
+    kp_source = model.kp_extractor(x['source'])
+
+    if driving_lr is not None:
+        kp_driving = model.kp_extractor(driving_lr)
+    else:
+        kp_driving = model.kp_extractor(x['driving'])
+
+    # quick warmup
+    for _ in range(10):
+        generated = model.generator(x['source'],
+                                    kp_source=kp_source,
+                                    kp_driving=kp_driving,
+                                    update_source=True,
+                                    driving_lr=driving_lr)
+
+    starter, ender = torch.cuda.Event(enable_timing=True), torch.cuda.Event(
+        enable_timing=True)
+
+    # Just 50 round measurement
+    total_time = 0
+    for i in range(50):
+        starter.record()
+        with torch.no_grad():
+            generated = model.generator(x['source'],
+                                        kp_source=kp_source,
+                                        kp_driving=kp_driving,
+                                        update_source=True,
+                                        driving_lr=driving_lr)
+        ender.record()
+        # WAIT FOR GPU SYNC
+        torch.cuda.synchronize()
+        curr_time = starter.elapsed_time(ender)
+        total_time += curr_time
+
+    return total_time / 50
+
+
+def get_gen_input(model=None, x=None):
+    """
+    Store the input of the model after this function is run once.
+    Is used for conveniently getting dummy inputs when I want to run a generator, without me needing to make sure sizes are set up correctly.
+    """
+    if not x is None:
+        driving_lr = x.get('driving_lr', None)
+
+        kp_source = model.kp_extractor(x['source'])
+
+        if driving_lr is not None:
+            kp_driving = model.kp_extractor(driving_lr)
+        else:
+            kp_driving = model.kp_extractor(x['driving'])
+
+        get_gen_input.inputs = (x['source'], kp_source, kp_driving, True,
+                                driving_lr)
+    return get_gen_input.inputs
+
+
+def calculate_macs(model, file_name=None):
+    """
+    Calculates the macs dict for the model.
+    Output is some sort of dictionary with key=layer, value = macs
+    """
+
+    inputs = get_gen_input()
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        result = profile_macs(model, inputs, reduction=None)
+    return result
+
+
+def total_macs(model):
+    """
+    Returns single number for total macs in model
+    """
+
+    macs_dict = calculate_macs(model)
+    return sum(macs_dict.values())
+
+def reverse_sort(input_list):
+    """
+    Input list is of the form: [(a, b), (c, d), ...]
+    We want to return the list sorted by the first element of each tuple
+    """
+    return sorted(input_list, key=lambda x: x[0], reverse=True)
+
+
+def f_set(weight, target, prune_indices):
```

Comment:
I think it would be very helpful if all node related functions were in one place/inside a class. This is really hard to follow right now